File: implementations/java/server/src/test/java/com/theshuai/tunnelserver/websocket/PublicTransferRelayFrameTests.java
package com.theshuai.tunnelserver.websocket;

import org.junit.jupiter.api.Test;

import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

class PublicTransferRelayFrameTests {
    @Test
    void decodesClientFrameAndAddsAuthenticatedSource() {
        byte[] app = appFrame(2, new byte[]{1, 2, 3});
        byte[] client = relayFrame("peer-b", "", app);

        PublicTransferRelayFrame.ClientFrame decoded =
                PublicTransferRelayFrame.decodeClient(ByteBuffer.wrap(client));

        assertEquals("peer-b", decoded.targetPeerId());
        assertEquals(2, decoded.appType());
        assertArrayEquals(app, decoded.appFrame());
        assertArrayEquals(relayFrame("peer-b", "peer-a", app),
                PublicTransferRelayFrame.encodeServer("peer-b", "peer-a", app));
    }

    @Test
    void rejectsSpoofedSourceAndTrailingAppBytes() {
        byte[] app = appFrame(1, new byte[]{4});
        assertThrows(IllegalArgumentException.class,
                () -> PublicTransferRelayFrame.decodeClient(
                        ByteBuffer.wrap(relayFrame("peer-b", "spoofed", app))));

        byte[] malformed = java.util.Arrays.copyOf(app, app.length + 1);
        assertThrows(IllegalArgumentException.class,
                () -> PublicTransferRelayFrame.decodeClient(
                        ByteBuffer.wrap(relayFrame("peer-b", "", malformed))));
    }

    private static byte[] relayFrame(String targetPeerId, String sourcePeerId, byte[] app) {
        byte[] target = targetPeerId.getBytes(StandardCharsets.UTF_8);
        byte[] source = sourcePeerId.getBytes(StandardCharsets.UTF_8);
        ByteBuffer frame = ByteBuffer.allocate(14 + target.length + source.length + app.length);
        frame.putInt(0x53545752).put((byte) 2).put((byte) 0);
        frame.putShort((short) target.length).putShort((short) source.length).putInt(app.length);
        frame.put(target).put(source).put(app);
        return frame.array();
    }

    private static byte[] appFrame(int type, byte[] payload) {
        ByteBuffer frame = ByteBuffer.allocate(72 + payload.length);
        frame.putInt(0x53544150).put((byte) 2).put((byte) type).putShort((short) 0);
        frame.put(new byte[16]);
        frame.putInt(0).putInt(1).putInt(payload.length).putInt(payload.length);
        frame.put(new byte[32]).put(payload);
        return frame.array();
    }
}

File: implementations/csharp/server/tests/Specus.IntegrationTests/PublicTransferRelayFrameTests.cs
using System.Buffers.Binary;
using System.Text;
using Specus.Server.WebSockets;

namespace Specus.IntegrationTests;

public sealed class PublicTransferRelayFrameTests
{
    [Fact]
    public void ClientFrameIsStrictAndServerAddsAuthenticatedSource()
    {
        var app = AppFrame(2, [1, 2, 3]);
        var client = RelayFrame("peer-b", string.Empty, app);

        var decoded = PublicTransferRelayFrame.DecodeClient(client);

        Assert.Equal("peer-b", decoded.TargetPeerId);
        Assert.Equal(2, decoded.AppType);
        Assert.Equal(app, decoded.AppFrame);
        Assert.Equal(RelayFrame("peer-b", "peer-a", app),
            PublicTransferRelayFrame.EncodeServer("peer-b", "peer-a", app));
    }

    [Fact]
    public void SpoofedSourceAndTrailingAppBytesAreRejected()
    {
        var app = AppFrame(1, [4]);
        Assert.Throws<ArgumentException>(() =>
            PublicTransferRelayFrame.DecodeClient(RelayFrame("peer-b", "spoofed", app)));
        Assert.Throws<ArgumentException>(() =>
            PublicTransferRelayFrame.DecodeClient(
                RelayFrame("peer-b", string.Empty, [.. app, 0])));
    }

    private static byte[] RelayFrame(string targetPeerId, string sourcePeerId, byte[] app)
    {
        var target = Encoding.UTF8.GetBytes(targetPeerId);
        var source = Encoding.UTF8.GetBytes(sourcePeerId);
        var result = new byte[14 + target.Length + source.Length + app.Length];
        "STWR"u8.CopyTo(result);
        result[4] = 2;
        BinaryPrimitives.WriteUInt16BigEndian(result.AsSpan(6, 2), (ushort)target.Length);
        BinaryPrimitives.WriteUInt16BigEndian(result.AsSpan(8, 2), (ushort)source.Length);
        BinaryPrimitives.WriteUInt32BigEndian(result.AsSpan(10, 4), (uint)app.Length);
        var offset = 14;
        target.CopyTo(result, offset);
        offset += target.Length;
        source.CopyTo(result, offset);
        offset += source.Length;
        app.CopyTo(result, offset);
        return result;
    }

    private static byte[] AppFrame(byte appType, byte[] payload)
    {
        var result = new byte[72 + payload.Length];
        "STAP"u8.CopyTo(result);
        result[4] = 2;
        result[5] = appType;
        BinaryPrimitives.WriteUInt32BigEndian(result.AsSpan(28, 4), 1);
        BinaryPrimitives.WriteUInt32BigEndian(result.AsSpan(32, 4), (uint)payload.Length);
        BinaryPrimitives.WriteUInt32BigEndian(result.AsSpan(36, 4), (uint)payload.Length);
        payload.CopyTo(result, 72);
        return result;
    }
}
